<h1><center>Udemy Courses Analysis<center>

<center>Analysis of data gathered from Udemy Courses. Link is available at https://www.kaggle.com/andrewmvd/udemy-courses?select=udemy_courses.csv<center>

### I. Importing modules and libraries

For this step, I imported numpy, and matplotlib modules. After setting up the modules, the csv file was also imported. 

In [1]:
import numpy  as np
import pandas as pd 

In [2]:
import os
import csv

os.chdir("C:\\Users\\maegr\Documents\Maymay\DataScience\modules")

#Opens the data file as df
with open("udemy_courses.csv", "r") as courses_df:
    courses_df = pd.read_csv('udemy_courses.csv')

### II. Describing the Courses

Instead of getting the statistical values separately, the **describe** syntax was used so that it would be easier to analyze the data.

In [3]:
courses_df.describe()

course_id        price  num_subscribers   num_reviews  num_lectures  \
count  3.678000e+03  3678.000000      3678.000000   3678.000000   3678.000000   
mean   6.759720e+05    66.049483      3197.150625    156.259108     40.108755   
std    3.432732e+05    61.005755      9504.117010    935.452044     50.383346   
min    8.324000e+03     0.000000         0.000000      0.000000      0.000000   
25%    4.076925e+05    20.000000       111.000000      4.000000     15.000000   
50%    6.879170e+05    45.000000       911.500000     18.000000     25.000000   
75%    9.613555e+05    95.000000      2546.000000     67.000000     45.750000   
max    1.282064e+06   200.000000    268923.000000  27445.000000    779.000000   

       content_duration  
count       3678.000000  
mean           4.094517  
std            6.053840  
min            0.000000  
25%            1.000000  
50%            2.000000  
75%            4.500000  
max           78.500000

#### Conclusions: 
1. If someone is planning to take a Udemy course, it must be noted that the average price of Udemy Courses is **66.049 USD**. Some of the courses are entirely **free**, while others could go as high as **200 USD**
2. The average number of subcsribers per course is **3197**
3. Some courses could have as high as **27455** reviews while some courses don't have any review at all.
3. In order to finish a course, one must spend an average of **4.09 hours**. 

### III. Amount of Courses by Subject

The pie chart below describes the distribution of the amount of courses per subject

In [4]:
courses_df['subject'].value_counts() / courses_df['subject'].value_counts().sum()

Web Development        0.326264
Business Finance       0.324905
Musical Instruments    0.184883
Graphic Design         0.163948
Name: subject, dtype: float64

#### Conclusions: 
1. About 32.6% of lessons from Udemy is about Web Development, which makes it the most popular subject in the site. 
2. The Web Development course is followed closely by Business Finance with around 32.5%. 
3. Musical Instruments is at third, while Graphic Design is placed at last.

### IV. Standard Deviation and Mean
Part III already discussed the following details but this is another way of getting the standard deviation

In [5]:
print("Mean Values")
print( "Price: " + str(courses_df["price"].mean()) )
print( "Number of subscribers: "+ str(courses_df["num_subscribers"].mean()) )
print( "Number of reviews: "+ str(courses_df["num_reviews"].mean()) )
print( "Number of lectures: "+ str(courses_df["num_lectures"].mean()) )
print( "Content Duration: "+ str(courses_df["content_duration"].mean()) )

print("\nStandard Deviation")
print( "Price: " + str(courses_df["price"].std()) )
print( "Number of subscribers: "+ str(courses_df["num_subscribers"].std()) )
print( "Number of reviews: "+ str(courses_df["num_reviews"].std()) )
print( "Number of lectures: "+ str(courses_df["num_lectures"].std()) )
print( "Content Duration: "+ str(courses_df["content_duration"].std()) )

Mean Values
Price: 66.0494834148994
Number of subscribers: 3197.150625339859
Number of reviews: 156.25910821098424
Number of lectures: 40.108754758020666
Content Duration: 4.0945169476164605

Standard Deviation
Price: 61.00575547648037
Number of subscribers: 9504.117010438013
Number of reviews: 935.452044243373
Number of lectures: 50.38334552355022
Content Duration: 6.053840414790095


#### Conclusions:
1. Based from the table above, we can infer that the there the standard deviation between the price and the number of lectures is small hence, they are more closely distributed around the mean value gathered earlier.

### V. Maximum and Minimum Values
Below are the maximum and minimum values of the given data

In [6]:
print( "Maximum Course Price: " + str(courses_df["price"].max()) )
print( "Minimum Course Price: " + str(courses_df["price"].min()) )

print( "\nMaximum Number of subscribers: "+ str(courses_df["num_subscribers"].max()) )
print( "Minimum Number of subscribers: "+ str(courses_df["num_subscribers"].min()) )

print( "\nMaximum Number of reviews: "+ str(courses_df["num_reviews"].max()) )
print( "Minimum Number of reviews: "+ str(courses_df["num_reviews"].min()) )

print( "\nMaximum Number of lectures: "+ str(courses_df["num_lectures"].max()) )
print( "Minimum Number of lectures: "+ str(courses_df["num_lectures"].min()) )

print( "\nMaximum Content Duration: "+ str(courses_df["content_duration"].max()) )
print( "Minimum Content Duration: "+ str(courses_df["content_duration"].min()) )

Maximum Course Price: 200
Minimum Course Price: 0

Maximum Number of subscribers: 268923
Minimum Number of subscribers: 0

Maximum Number of reviews: 27445
Minimum Number of reviews: 0

Maximum Number of lectures: 779
Minimum Number of lectures: 0

Maximum Content Duration: 78.5
Minimum Content Duration: 0.0


Conclusions:
1. The most expensive course costs 200USD while there are courses that are free.
2. The most subscribed course has 268923 subscribers
3. The most reviewed course has 27445 reviews
4. The most number of lectures in a course is 779. Surprisingly, there are courses that do not have any lectures at all.
5. The maximum content duration of a course is 78.5 hours, while again, there are courses that have 0 content duration.

### VI. Most Popular Paid Courses
The table below shows the most popular paid courses in Udemy based on the number of subscribers

In [7]:
paid_courses_df = courses_df.query("price != 0")
top25_paid = paid_courses_df.sort_values("num_subscribers", ascending=False)[0:25].sort_values("num_subscribers", ascending=False)

top25_paid

course_id                                       course_title  \
3230     625204                         The Web Developer Bootcamp   
3232     764164              The Complete Web Developer Course 2.0   
2619      94430             Learn Javascript & JQuery From Scratch   
3247     364426          JavaScript: Understanding the Weird Parts   
1979     238934  Pianoforall - Incredible New Way To Learn Pian...   
3204     756150  Angular 4 (formerly Angular 2) - The Complete ...   
2701      11174                Become a Web Developer from Scratch   
3246     289230                     Learn and Understand AngularJS   
3251     461160                        Learn and Understand NodeJS   
2662     556248  The Complete HTML & CSS Course - From Novice T...   
3175     580466  Complete PHP Course With Bootstrap3 CMS System...   
3205      80940        Build Websites from Scratch with HTML & CSS   
3254     705264                            Modern React with Redux   
2479      19603  Learning Dynamic Website Design - PHP MySQL an...   
3249     437398  Build Responsive Real World Websites with HTML...   
2478      28295  Learn Web Designing & HTML5/CSS3 Essentials in...   
2561      28295  Learn Web Designing & HTML5/CSS3 Essentials in...   
3228     719002  Angular 2 with TypeScript for Beginners: The P...   
2806     674608  The Complete Web Developer Masterclass: Beginn...   
1213     820194  Photoshop for Entrepreneurs - Design 11 Practi...   
2497      16646                        Web Programming with Python   
3233     446134  Ultimate Web Designer & Developer Course: Buil...   
3208     519442        The Complete Ruby on Rails Developer Course   
1531     249126                            Logo Design Essentials    
1982      42271      Complete Guitar System - Beginner to Advanced   

                                                    url  is_paid  price  \
3230  https://www.udemy.com/the-web-developer-bootcamp/     True    200   
3232  https://www.udemy.com/the-complete-web-develop...     True    200   
2619  https://www.udemy.com/learn-javascript-jquery-...     True     30   
3247       https://www.udemy.com/understand-javascript/     True    175   
1979  https://www.udemy.com/pianoforall-incredible-n...     True    200   
3204  https://www.udemy.com/the-complete-guide-to-an...     True    190   
2701  https://www.udemy.com/how-to-become-a-web-deve...     True    120   
3246             https://www.udemy.com/learn-angularjs/     True    175   
3251           https://www.udemy.com/understand-nodejs/     True    195   
2662                 https://www.udemy.com/htmlwebsite/     True    195   
3175         https://www.udemy.com/complete-php-course/     True    195   
3205       https://www.udemy.com/build-website-scratch/     True    145   
3254                 https://www.udemy.com/react-redux/     True    180   
2479  https://www.udemy.com/learning-dynamic-website...     True     50   
3249  https://www.udemy.com/design-and-develop-a-kil...     True    195   
2478  https://www.udemy.com/build-beautiful-html5-we...     True     75   
2561  https://www.udemy.com/build-beautiful-html5-we...     True     75   
3228  https://www.udemy.com/angular-2-tutorial-for-b...     True    150   
2806          https://www.udemy.com/webdevelopercourse/     True    195   
1213  https://www.udemy.com/photoshop-for-entreprene...     True    200   
2497  https://www.udemy.com/web-programming-with-pyt...     True     50   
3233        https://www.udemy.com/web-developer-course/     True    200   
3208  https://www.udemy.com/the-complete-ruby-on-rai...     True    195   
1531                 https://www.udemy.com/logo-design/     True     20   
1982  https://www.udemy.com/complete-guitar-system-b...     True    145   

      num_subscribers  num_reviews  num_lectures           level  \
3230           121584        27445           342      All Levels   
3232           114512        22412           304      All Levels   
2619            84897         

#### Conclusions:
1. The most popular paid course is "The Web Developer Bootcamp". 
2. 21 out of 25 most popular paid courses is about Web Development. From this and the conclusion made earlier, it is possible that many udemy users are interested in learning about Web Development.
3. Those most popular paid course in musical instruments is about learning to play piano and guitar.

### VII Most Popular Free Courses

In [8]:
free_courses_df = courses_df.query("price == 0")
top25_free = free_courses_df.sort_values("num_subscribers", ascending=False)[0:25].sort_values("num_subscribers", ascending=False)

top25_free

course_id                                       course_title  \
2827      41295               Learn HTML5 Programming From Scratch   
3032      59014                     Coding for Entrepreneurs Basic   
2783     173548  Build Your First Website in 1 Week with HTML5 ...   
1896      19421              Free Beginner Electric Guitar Lessons   
2589     473160  Web Design for Web Developers: Build Beautiful...   
3289     130064  Practical PHP: Master the Basics and Code Dyna...   
3665      21386               Beginner Photoshop to HTML5 and CSS3   
2782      65330  Web Development By Doing: HTML / CSS From Scratch   
3325     405926  HTML and CSS for Beginners - Build a Website &...   
492       49798  Bitcoin or How I Learned to Stop Worrying and ...   
2592     314462                               Quickstart AngularJS   
3219     128946      Learn Responsive Web Development from Scratch   
143       48841    Accounting in 60 Minutes - A Brief Introduction   
1461     348116  Photoshop In-Depth: Master all of Photoshop's ...   
103      133536               Stock Market Investing for Beginners   
3291      94432                Learn PHP Fundamentals From Scratch   
1286      17349  Figure Drawing From Life Using The Reilly Tech...   
2004      41884                Getting Started with Playing Guitar   
3385     707962                              Javascript Essentials   
1390     399938      Professional Logo Design in Adobe Illustrator   
3328     708578  Bootstrap 4 Quick Start: Code Modern Responsiv...   
3238     772618  Learn Object Oriented PHP By Building a Comple...   
2779     294794          Create An HTML5 Video Player From Scratch   
3365     511934                            AngularJS For Beginners   
2649     267560  Learn E-Commerce Website in PHP & MySQL From S...   

                                                    url  is_paid  price  \
2827  https://www.udemy.com/learn-html5-programming-...    False      0   
3032  https://www.udemy.com/coding-for-entrepreneurs...    False      0   
2783  https://www.udemy.com/build-your-first-website...    False      0   
1896  https://www.udemy.com/beginner-electric-guitar...    False      0   
2589          https://www.udemy.com/web-design-secrets/    False      0   
3289       https://www.udemy.com/code-dynamic-websites/    False      0   
3665              https://www.udemy.com/psd-html5-css3/    False      0   
2782  https://www.udemy.com/web-development-learn-by...    False      0   
3325  https://www.udemy.com/html-and-css-for-beginne...    False      0   
492   https://www.udemy.com/bitcoin-or-how-i-learned...    False      0   
2592        https://www.udemy.com/quickstart-angularjs/    False      0   
3219  https://www.udemy.com/learn-responsive-web-dev...    False      0   
143   https://www.udemy.com/accounting-in-60-minutes...    False      0   
1461             https://www.udemy.com/photoshop-tools/    False      0   
103   https://www.udemy.com/the-beginners-guide-to-t...    False      0   
3291  https://www.udemy.com/learn-php-fundamentals-f...    False      0   
1286  https://www.udemy.com/figure-drawing-from-life...    False      0   
2004  https://www.udemy.com/getting-started-with-pla...    False      0   
3385       https://www.udemy.com/javascript-essentials/    False      0   
1390  https://www.udemy.com/professional-logo-design...    False      0   
3328                 https://www.udemy.com/bootstrap-4/    False      0   
3238  https://www.udemy.com/learn-object-oriented-ph...    False      0   
2779          https://www.udemy.com/html5-video-player/    False      0   
3365  https://www.udemy.com/angularjs-for-beginners-...    False      0   
2649  https://www.udemy.com/ecommerce-website-in-php...    False      0   

      num_subscribers  num_reviews  num_lectures           level  \
2827           268923         8629            45      All Levels   
3032           161029          279            27  Beginner Level   
2783           120291         

#### Conclusions:
1. The most popular free course is "Learn HTML5 Programming from scratch". 
2. 21 out of 25 most popular free courses is about Web Development. 
3. Those most popular free course in musical instruments is about learning to play electric guitar, titled "Free Beginner Electric Guitar Lessons"

### VIII. Highest reviewed course

In [9]:
num_reviews_df = courses_df.query("num_reviews != 0")
top25_reviewed = num_reviews_df.sort_values("num_reviews", ascending=False)[0:25].sort_values("num_reviews", ascending=True).reset_index(drop=True).reset_index()
top25_reviewed.max()

index                                                                 24
course_id                                                         995016
course_title           Web Design for Web Developers: Build Beautiful...
url                          https://www.udemy.com/web-developer-course/
is_paid                                                             True
price                                                                200
num_subscribers                                                   268923
num_reviews                                                        27445
num_lectures                                                         362
level                                                 Intermediate Level
content_duration                                                    43.0
published_timestamp                                 2016-11-08T18:55:21Z
subject                                                  Web Development
dtype: object

#### Conclusions:
1. The highest reviewed course is about Web Designing under the subject Web Development. 
2. At the time of the csv file upload, the title of the course is "Web Design for Web Developers" but currently, it is titled as "Ultimate Web Designer & Web Developer Course for 2021" 
3. It takes 43 hours to finish the course. 
4. It costs around 200 USD before but currently, it is priced at 89.99 USD.
5. The course is good for Intermediate Level learners

To check the course, visit the link https://www.udemy.com/web-developer-course/

## General Conclusions

IF someone is interested in learning about Web Development, Business Finance, Musical Instruments, and Graphics Design, it is recommended that they visit Udemy since there's a variety of courses they could try from the website. There are also free courses available. Meanwhile, if they don't mind paying for the contents, they should know that the average price per course is 66 USD and could possibly go as high as 200 USD.